### PR TITLE
Create Typescript Definitions

### PR DESCRIPTION
**Pull request by franklixuefei**
Create Typescript Definitions.
There already exists one under `@types/`, but that one was not updated for a while and does not match the implementation. It also doesn't include types for `PENDING`/`FULFILLED`/`REJECTED` consts.

### CODE REVIEW

**Comment by pburtchaell** [comment]
Thanks! I don't use TS, so I forget to update it when changes are made.

Is there some test(s) we can add to make sure the TS definitions are always up to date?

**Comment by franklixuefei** [comment]
@pburtchaell Currently I'm not very sure how we can make sure the types are up to date. However, I think we can do `import promiseMiddleware from 'redux-promise-middleware '` in a test file written in typescript (with`.ts` extension) and then write some unit tests in there just to consume `promiseMiddleware` and do some dummy things, just to make sure the type works. However, that involves pulling in the whole typescript pipeline I guess.

**Comment by franklixuefei** [comment]
@pburtchaell But I think for now you are good, because the interface for this project is pretty simple - 3 `const`s export and a `function` export. So maybe it's not worth the effort for now.

**Comment by franklixuefei** [comment]
@pburtchaell One way (and probably the ideal way) to resolve and avoid having any issues with type definition files is to write a type definition file generator that parses some documentation, like `.md` files, in which you define the types for each component/function/variable you want to export, and then, with each release, the type definition files can be dynamically generated and kept up to date with the actual implementation, as long as you also remember to keep the documentation up to date.
I came across this idea after I looked at this repo([DevExpress Reactive](https://github.com/DevExpress/devextreme-reactive)), where they generate type definition files (`.d.ts` files) using a [script](https://github.com/DevExpress/devextreme-reactive/blob/master/scripts/ts-generator.js).
Just some thoughts.

**Comment by pburtchaell** [comment]
Thanks @franklixuefei. I'm a locked down with a project this weekend and for most of next week, but I'll take a look at this as soon as I can.

**Comment by pburtchaell** [comment]
Here's what I think: merge this PR and just add a comment to `index.js` reminding me and contributors to also update `index.ts`. Like you said, since the interface is simple, it's not worth the time it would take to add TS tests or a file generator.

**Comment by franklixuefei** [comment]
@pburtchaell I agree. I'll add the comment in `index.js` soon.

**Comment by franklixuefei** [comment]
@pburtchaell Added contributor note in `index.js`, as well as mentioning TypeScript support in `Introduction.md`.

**Comment by pburtchaell** [comment]
@franklixuefei Seems so. I'll take a look at it and bump the version.

**Comment by pburtchaell** [comment]
@franklixuefei I'm going to get to this ASAP, but I'm a bit locked down with priorities from another project—it's on my radar though.

**Comment by franklixuefei** [comment]
@pburtchaell Thanks for the update! Take your time :-D

File: src/index.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
 import isPromise from './isPromise.js';
 
+/**
+ * Note to contributors: Please also remember to check and make sure that `index.d.ts` is also up to date with
```
Comment (pburtchaell) [inline src/index.js]: This line is breaking the build: just need to break it so the character count isn't over 100. 

Build error: https://travis-ci.org/pburtchaell/redux-promise-middleware/jobs/367749777#L493

**Comment by pburtchaell** [comment]
Turns out the build wasn't failing due to the Node version (but it needed to be updated anyway). When you can a chance, can you resolve the change I requested?

**Comment by franklixuefei** [comment]
@pburtchaell test passed.

**Comment by pburtchaell** [comment]
I forgot to mention... this was released in `redux-promise-middleware@5.1.1`: https://github.com/pburtchaell/redux-promise-middleware/releases/tag/5.1.1

**Comment by pburtchaell** [comment]
Thanks @franklixuefei!

**Comment by paulkoegel-wessels** [comment]
@pburtchaell shouldn't the type definitions in the definitelytyped repo be removed then?

I just upgraded our app to Redux 4 and stumbled upon `@types/redux-promise-middleware` still requiring redux 3.6.

There's already an issue on definitelytyped for removing deprecated packages:
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/13364

**Comment by pburtchaell** [comment]
Yes, they should then. @paulkoegel-wessels, can you open an issue here to track this?

**Pull request merged**